contracts/Governance.sol

```diff
@@ -211,23 +213,152 @@ contract Governance is Context, Ownable, ERC165, EIP712 {
         return _votingPeriod;
     }
 
-    function setTreasury(address treasury_) external virtual onlyOwner {
+    function setInitialTreasury(address treasury_) external virtual onlyOwner {
+        require(
+            treasury == address(0),
+            'Governance::setInitialTreasury: treasury was set'
+        );
         treasury = treasury_;
+        protocolsRoles[treasury_] = TREASURY_ROLE;
     }
 
-    function setVotesOracle(address votesOracle_) external virtual onlyOwner {
+    function setTreasury(address treasury_) external virtual onlyGovernance {
+        require(
+            treasury != address(0) && treasury != treasury_,
+            "Governance::setTreasury: treasury wasn't set or the same"
+        );
+        treasury = treasury_;
+        protocolsRoles[treasury_] = TREASURY_ROLE;
+    }
+
+    function setVotesOracle(address votesOracle_)
+        external
+        virtual
+        onlyGovernance
+    {
         openVotingOracle = IOpenVoting(votesOracle_);
         roleVotingOracle = IRoleVoting(votesOracle_);
     }
 
-    function setVotingDelay(uint256 votingDelay_) external virtual onlyOwner {
+    function setVotingDelay(uint256 votingDelay_)
+        external
+        virtual
+        onlyGovernance
+    {
         _votingDelay = votingDelay_;
     }
 
-    function setVotingPeriod(uint256 votingPeriod_) external virtual onlyOwner {
+    function setVotingPeriod(uint256 votingPeriod_)
+        external
+        virtual
+        onlyGovernance
+    {
         _votingPeriod = votingPeriod_;
     }
 
+    function registerProtocol(
+        address protocol,
+        bytes32 role,
+        uint256 quorum
+    ) external virtual onlyGovernance {
+        protocolsRoles[protocol] = role;
+        protocolsQuorums[protocol] = quorum;
+    }
+
+    function unregisterProtocol(address protocol)
+        external
+        virtual
+        onlyGovernance
+    {
+        delete protocolsRoles[protocol];
+        protocolsQuorums[protocol] = 0;
+    }
+
+    function registerRole(bytes32 role) external virtual onlyGovernance {
+        rolesList.push(role);
+        _roles[role] = rolesList.length;
+    }
+
+    function unregisterRole(bytes32 role) external virtual onlyGovernance {
+        rolesList[_roles[role]] = rolesList[rolesList.length - 1];
+        _roles[rolesList[rolesList.length - 1]] = _roles[role];
+        delete rolesList[rolesList.length - 1];
```

Comment:
If I remember correctly:
delete array will not change array length, it will reset the value at that index to be the default value
so you will have to follow by "rolesList.length--;"
Or you can just "rolesList.pop()" now the last element is gone
Can you double check?

Reply:
> delete array will not change array length, 

I believe you are correct. `delete` zero's the element contents without shortening the size of the array (that array index is now empty).